# MATH 103.1 Exam 3 - Transforming texts to numbers
Name: Yumol, Dianne


**TA**: Load the data set `news_data.csv` and save it to `news_data`.

In [1]:
import pandas as pd
news_data = pd.read_csv("news_data.csv")
news_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11314 entries, 0 to 11313
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Text      11314 non-null  object
 1   Category  11314 non-null  object
dtypes: object(2)
memory usage: 176.9+ KB


## Problem 1. (10 pts)

A sample is a row in the data frame. 

### 1.1 How many samples are there in `news_data`?  

**Answer**: 11314

### 1.2  How many columns?  

**Answer**: 2

### 1.3 What are the column names?  

**Answer**: Text and Category

### 1.4 Are there missing values?  

**Answer**: No, there are 0 null values.



## Problem 2. (10 pts)

Each sample in `news_data` is labelled according to Category. 

### 2.1 Create a pandas Series which is a frequency table of the `Category` column in `news_data`.   

In [4]:
## Answer to 2.1 
categories_count = news_data.value_counts("Category")
print(categories_count)

Category
rec.sport.hockey            600
soc.religion.christian      599
rec.motorcycles             598
rec.sport.baseball          597
sci.crypt                   595
rec.autos                   594
sci.med                     594
comp.windows.x              593
sci.space                   593
sci.electronics             591
comp.os.ms-windows.misc     591
comp.sys.ibm.pc.hardware    590
misc.forsale                585
comp.graphics               584
comp.sys.mac.hardware       578
talk.politics.mideast       564
talk.politics.guns          546
alt.atheism                 480
talk.politics.misc          465
talk.religion.misc          377
dtype: int64


### 2.2 View some basic statistical details like percentile, mean, std, etc. of `categories_count`?   

In [5]:
# Answer to 2.2
import numpy as np
categories_count.describe()

count     20.000000
mean     565.700000
std       58.251813
min      377.000000
25%      574.500000
50%      591.000000
75%      594.250000
max      600.000000
dtype: float64

### 2.3 Which category has the highest count? Which has the lowest count?

**Answer**: rec.sport.hockey has the highest count (600) while talk.religion.misc has the smallest count (377)


## Problem 3.  Use the `categories_count` in *Problem 2*. (20 points)

### 3.1 Create a pandas DataFrame from `categories_count` with two columns: `Category` and `Count`. 

In [6]:
# Answer to 3.1
categories_df = pd.DataFrame({"Category": categories_count.index, "Count": categories_count.values})
categories_df = categories_df.reset_index(drop=True)
print(categories_df)

                    Category  Count
0           rec.sport.hockey    600
1     soc.religion.christian    599
2            rec.motorcycles    598
3         rec.sport.baseball    597
4                  sci.crypt    595
5                  rec.autos    594
6                    sci.med    594
7             comp.windows.x    593
8                  sci.space    593
9            sci.electronics    591
10   comp.os.ms-windows.misc    591
11  comp.sys.ibm.pc.hardware    590
12              misc.forsale    585
13             comp.graphics    584
14     comp.sys.mac.hardware    578
15     talk.politics.mideast    564
16        talk.politics.guns    546
17               alt.atheism    480
18        talk.politics.misc    465
19        talk.religion.misc    377


### 3.2 Create a new column called `Branch` in `categories_df`.  

Its elements are the respective substring of the elements consisting of characters before the `.` symbol in `Category` column. For example, since the category in the first row is `rec.sport.hockey`, then the first element of `Branch` is `rec`.


In [7]:
# Answer to 3.2
categories_df["Branch"] = categories_df["Category"].str.split(".", expand=True)[0]
print(categories_df)

                    Category  Count Branch
0           rec.sport.hockey    600    rec
1     soc.religion.christian    599    soc
2            rec.motorcycles    598    rec
3         rec.sport.baseball    597    rec
4                  sci.crypt    595    sci
5                  rec.autos    594    rec
6                    sci.med    594    sci
7             comp.windows.x    593   comp
8                  sci.space    593    sci
9            sci.electronics    591    sci
10   comp.os.ms-windows.misc    591   comp
11  comp.sys.ibm.pc.hardware    590   comp
12              misc.forsale    585   misc
13             comp.graphics    584   comp
14     comp.sys.mac.hardware    578   comp
15     talk.politics.mideast    564   talk
16        talk.politics.guns    546   talk
17               alt.atheism    480    alt
18        talk.politics.misc    465   talk
19        talk.religion.misc    377   talk


### 3.3 What are the unique branches and how many in `categories_df`?  

In [8]:
# Answer to 3.3
print("unique branches:", categories_df["Branch"].unique())
print("number of branches:", len(categories_df["Branch"].unique()))

unique branches: ['rec' 'soc' 'sci' 'comp' 'misc' 'talk' 'alt']
number of branches: 7


### 3.4 Sort `categories_df` with respect to `Branch` in ascending order followed by sorting `Count` in descending order.  


In [9]:
categories_df = categories_df.sort_values(["Branch", "Count"], ascending=[True, False])
print(categories_df)

                    Category  Count Branch
17               alt.atheism    480    alt
7             comp.windows.x    593   comp
10   comp.os.ms-windows.misc    591   comp
11  comp.sys.ibm.pc.hardware    590   comp
13             comp.graphics    584   comp
14     comp.sys.mac.hardware    578   comp
12              misc.forsale    585   misc
0           rec.sport.hockey    600    rec
2            rec.motorcycles    598    rec
3         rec.sport.baseball    597    rec
5                  rec.autos    594    rec
4                  sci.crypt    595    sci
6                    sci.med    594    sci
8                  sci.space    593    sci
9            sci.electronics    591    sci
1     soc.religion.christian    599    soc
15     talk.politics.mideast    564   talk
16        talk.politics.guns    546   talk
18        talk.politics.misc    465   talk
19        talk.religion.misc    377   talk


## Problem 4 Create a subset of `news_data` that are `comp`s in the `Branch` column. (20 points)

### 4.1 Find the categories that belong to the `comp` branch.  

In [10]:
# Answer to 4.1
comp_branch = categories_df[categories_df.Branch == "comp"]["Category"]
print(comp_branch)

7               comp.windows.x
10     comp.os.ms-windows.misc
11    comp.sys.ibm.pc.hardware
13               comp.graphics
14       comp.sys.mac.hardware
Name: Category, dtype: object


### Problem 4.2  Create a subset of `news_data` whose samples are in the `comp_branch`.  

In [11]:
# Answer to 4.2
comp_news = news_data[news_data['Category'].isin(comp_branch)]
comp_news.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2936 entries, 1 to 11312
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Text      2936 non-null   object
 1   Category  2936 non-null   object
dtypes: object(2)
memory usage: 68.8+ KB


### 4.3 Create a frequency table from the `Category` column of `comp_news`.  

In [12]:
# Answer to 4.3
print("\nFrequency Table: ", comp_news.value_counts("Category"))


Frequency Table:  Category
comp.windows.x              593
comp.os.ms-windows.misc     591
comp.sys.ibm.pc.hardware    590
comp.graphics               584
comp.sys.mac.hardware       578
dtype: int64


### 4.4 What is the shape of `comp_news` and print its  first five samples. 

In [13]:
#print("shape of comp_news: ", 
nrows, ncols = comp_news.shape
print(f"num_rows = {nrows}, num_cols = {ncols}")
print(f"\nFirst five samples:\n {comp_news.head()}")

num_rows = 2936, num_cols = 2

First five samples:
                                                 Text                  Category
1  From: guykuo@carson.u.washington.edu (Guy Kuo)...     comp.sys.mac.hardware
2  From: twillis@ec.ecn.purdue.edu (Thomas E Will...     comp.sys.mac.hardware
3  From: jgreen@amber (Joe Green)\nSubject: Re: W...             comp.graphics
7  From: bgrubb@dante.nmsu.edu (GRUBB)\nSubject: ...  comp.sys.ibm.pc.hardware
8  From: holmes7000@iscsvax.uni.edu\nSubject: WIn...   comp.os.ms-windows.misc


## Problem 5. Clean, lowering case,  tokenize, remove stopwords, and lemmatize the dataset (20 points)

In [14]:
import re
import nltk
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('omw-1.4')
nltk.download('stopwords')
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

[nltk_data] Downloading package punkt to C:\Users\Don
[nltk_data]     Bosco\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.
[nltk_data] Downloading package wordnet to C:\Users\Don
[nltk_data]     Bosco\AppData\Roaming\nltk_data...
[nltk_data] Downloading package omw-1.4 to C:\Users\Don
[nltk_data]     Bosco\AppData\Roaming\nltk_data...
[nltk_data] Downloading package stopwords to C:\Users\Don
[nltk_data]     Bosco\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


### 5.1 Remove punctuations and lower the case in text. 

In [15]:
no_punct_lowercase_news = [re.sub(r'[^\w\s]', '', text.lower()) for text in  comp_news.Text]

print(f"no_punct_lowercase_news length: {len(no_punct_lowercase_news)}")
print(f"\n first no_punct_lowercase_news: \n {no_punct_lowercase_news[0]}")

no_punct_lowercase_news length: 2936

 first no_punct_lowercase_news: 
 from guykuocarsonuwashingtonedu guy kuo
subject si clock poll  final call
summary final call for si clock reports
keywords siaccelerationclockupgrade
articleid shelley1qvfo9innc3s
organization university of washington
lines 11
nntppostinghost carsonuwashingtonedu

a fair number of brave souls who upgraded their si clock oscillator have
shared their experiences for this poll please send a brief message detailing
your experiences with the procedure top speed attained cpu rated speed
add on cards and adapters heat sinks hour of usage per day floppy disk
functionality with 800 and 14 m floppies are especially requested

i will be summarizing in the next two days so please add to the network
knowledge base if you have done the clock upgrade and havent answered this
poll thanks

guy kuo guykuouwashingtonedu



### 5.2 Tokenize the text in `no_punct_lowercase_news`.

In [16]:
# Answer to 5.2
tokenized_news = [word_tokenize(text) for text in no_punct_lowercase_news]
print(f"tokenized_news length: {len(tokenized_news)}")
print(f"\n first tokenized news: \n {tokenized_news[0]}")

tokenized_news length: 2936

 first tokenized news: 
 ['from', 'guykuocarsonuwashingtonedu', 'guy', 'kuo', 'subject', 'si', 'clock', 'poll', 'final', 'call', 'summary', 'final', 'call', 'for', 'si', 'clock', 'reports', 'keywords', 'siaccelerationclockupgrade', 'articleid', 'shelley1qvfo9innc3s', 'organization', 'university', 'of', 'washington', 'lines', '11', 'nntppostinghost', 'carsonuwashingtonedu', 'a', 'fair', 'number', 'of', 'brave', 'souls', 'who', 'upgraded', 'their', 'si', 'clock', 'oscillator', 'have', 'shared', 'their', 'experiences', 'for', 'this', 'poll', 'please', 'send', 'a', 'brief', 'message', 'detailing', 'your', 'experiences', 'with', 'the', 'procedure', 'top', 'speed', 'attained', 'cpu', 'rated', 'speed', 'add', 'on', 'cards', 'and', 'adapters', 'heat', 'sinks', 'hour', 'of', 'usage', 'per', 'day', 'floppy', 'disk', 'functionality', 'with', '800', 'and', '14', 'm', 'floppies', 'are', 'especially', 'requested', 'i', 'will', 'be', 'summarizing', 'in', 'the', 'next', 't

### 5.3 Remove the stop words in the `tokenized_news`. 

In [17]:
# Answer to 5.3
set_stop_words = set(stopwords.words('english'))
cleaned_tokenized_news = []
for tokens in tokenized_news:
    cleaned_tokenized_news.append([token for token in tokens if token not in set_stop_words])

print("cleaned_tokenized_news length: ", len(cleaned_tokenized_news))
print("\n first cleaned and tokenized news: \n", cleaned_tokenized_news[0])

cleaned_tokenized_news length:  2936

 first cleaned and tokenized news: 
 ['guykuocarsonuwashingtonedu', 'guy', 'kuo', 'subject', 'si', 'clock', 'poll', 'final', 'call', 'summary', 'final', 'call', 'si', 'clock', 'reports', 'keywords', 'siaccelerationclockupgrade', 'articleid', 'shelley1qvfo9innc3s', 'organization', 'university', 'washington', 'lines', '11', 'nntppostinghost', 'carsonuwashingtonedu', 'fair', 'number', 'brave', 'souls', 'upgraded', 'si', 'clock', 'oscillator', 'shared', 'experiences', 'poll', 'please', 'send', 'brief', 'message', 'detailing', 'experiences', 'procedure', 'top', 'speed', 'attained', 'cpu', 'rated', 'speed', 'add', 'cards', 'adapters', 'heat', 'sinks', 'hour', 'usage', 'per', 'day', 'floppy', 'disk', 'functionality', '800', '14', 'floppies', 'especially', 'requested', 'summarizing', 'next', 'two', 'days', 'please', 'add', 'network', 'knowledge', 'base', 'done', 'clock', 'upgrade', 'havent', 'answered', 'poll', 'thanks', 'guy', 'kuo', 'guykuouwashingtonedu

### 5.4 Lemmatization

In [20]:
lemmatizer = WordNetLemmatizer()
lemma_news = []
for tokens in cleaned_tokenized_news:
    lc_news = [lemmatizer.lemmatize(token, pos="v") for token in tokens]
    lc_news = [lemmatizer.lemmatize(token, pos="n") for token in lc_news]
    lc_news = [lemmatizer.lemmatize(token, pos="a") for token in lc_news]
    lc_news = [lemmatizer.lemmatize(token, pos="r") for token in lc_news]
    lc_news = [lemmatizer.lemmatize(token, pos="s") for token in lc_news]
    lemma_news.append(lc_news)

print("lemma_news length: ", len(lemma_news))
print("\n first lemma_news: \n", lemma_news[0])

lemma_news length:  2936

 first lemma_news: 
 ['guykuocarsonuwashingtonedu', 'guy', 'kuo', 'subject', 'si', 'clock', 'poll', 'final', 'call', 'summary', 'final', 'call', 'si', 'clock', 'report', 'keywords', 'siaccelerationclockupgrade', 'articleid', 'shelley1qvfo9innc3s', 'organization', 'university', 'washington', 'line', '11', 'nntppostinghost', 'carsonuwashingtonedu', 'fair', 'number', 'brave', 'soul', 'upgrade', 'si', 'clock', 'oscillator', 'share', 'experience', 'poll', 'please', 'send', 'brief', 'message', 'detail', 'experience', 'procedure', 'top', 'speed', 'attain', 'cpu', 'rat', 'speed', 'add', 'card', 'adapter', 'heat', 'sink', 'hour', 'usage', 'per', 'day', 'floppy', 'disk', 'functionality', '800', '14', 'floppy', 'especially', 'request', 'summarize', 'next', 'two', 'day', 'please', 'add', 'network', 'knowledge', 'base', 'do', 'clock', 'upgrade', 'havent', 'answer', 'poll', 'thank', 'guy', 'kuo', 'guykuouwashingtonedu']


## Problem 6.  `CountVectorizer` (20 points)

### 6.1 Create a function which will serve as a value to the `analyzer` parameter of `CountVectorizer`.

In [21]:
def text_analyzer(text):
    set_stop_words = set(stopwords.words('english'))
    lemmatizer = WordNetLemmatizer()

    no_punct_lowercase_text = re.sub(r'[^\w\s]', '', text.lower())
    tokenized_text = word_tokenize(no_punct_lowercase_text)
    no_stopwords = [token for token in tokenized_text if token not in set_stop_words]
    lc_text = [lemmatizer.lemmatize(token, pos="v") for token in no_stopwords]
    lc_text = [lemmatizer.lemmatize(token, pos="n") for token in lc_text]
    lc_text = [lemmatizer.lemmatize(token, pos="a") for token in lc_text]
    lc_text = [lemmatizer.lemmatize(token, pos="r") for token in lc_text]
    lc_text = [lemmatizer.lemmatize(token, pos="s") for token in lc_text]
    return lc_text

### 6.2 Create an instance of `CountVectorizer` 

In [22]:
# Answer to 6.2
from sklearn.feature_extraction.text import CountVectorizer

# Create an instance of CountVectorizer
news_vectorizer = CountVectorizer(analyzer=text_analyzer)

### 6.3 Fit and transform `news_vectorizer`.

In [23]:
# Answer to 6.3
# Fit and transform the vectorizer to the `comp_news.Text` data
news_tokens_bag = news_vectorizer.fit_transform(comp_news.Text)
news_tokens_bag

<2936x55472 sparse matrix of type '<class 'numpy.int64'>'
	with 271491 stored elements in Compressed Sparse Row format>

In [24]:
x = news_tokens_bag.nnz / (news_tokens_bag.shape[0] * news_tokens_bag.shape[1])
print(f"density = {format(x,'.2%')}")

density = 0.17%


### 6.4 Extract column names of the sparse matrix (`news_tokens_bag`) and save them to `features`. There are how many elements in `features`? Print 20 `features`' elements starting from index 29200.¶

In [26]:
features = news_vectorizer.get_feature_names_out()

# The number of features
print(len(features))

# Twenty elements starting from index 29200
print(features[29200:29220])

55472
['matqqs4u7v_orq7uqs4uslqrs2pl7qs4u' 'matrix' 'matrixvector' 'matrox'
 'matsbredelludacuuse' 'matsumura' 'matt' 'matt1ry69a1e967f4c8gsejx9kka4'
 'matt_harropmagicbbscorpapplecom' 'mattcenterlinecom' 'matteo' 'matter'
 'matthew' 'matthew_j_wilsonmcontentapanaorgau'
 'matthewalchemytncornelledu' 'matthewdabeefdesedu' 'matthias' 'mattias'
 'mattix' 'mattksuksuedu']
